pkg/cmd/publish-provisional-artifacts/main.go

```diff
@@ -36,47 +32,49 @@ const (
 
 var provisionalReleasePrefixRE = regexp.MustCompile(`^provisional_[0-9]{12}_`)
 
-type s3I interface {
-	GetObject(*s3.GetObjectInput) (*s3.GetObjectOutput, error)
-	PutObject(*s3.PutObjectInput) (*s3.PutObjectOutput, error)
-}
-
-func makeS3() (s3I, error) {
-	sess, err := session.NewSession(&aws.Config{
-		Region: aws.String("us-east-1"),
-	})
-	if err != nil {
-		return nil, err
-	}
-	return s3.New(sess), nil
-}
-
-var isReleaseF = flag.Bool("release", false, "build in release mode instead of bleeding-edge mode")
-var destBucket = flag.String("bucket", "", "override default bucket")
-var doProvisionalF = flag.Bool("provisional", false, "publish provisional binaries")
-var doBlessF = flag.Bool("bless", false, "bless provisional binaries")
-
-var (
-	// TODO(tamird,benesch,bdarnell): make "latest" a website-redirect
-	// rather than a full key. This means that the actual artifact will no
-	// longer be named "-latest".
-	latestStr = "latest"
-)
-
 func main() {
+	var isReleaseF = flag.Bool("release", false, "build in release mode instead of bleeding-edge mode")
+	var destBucket = flag.String("bucket", "", "override default bucket")
+	var gcsBucket = flag.String("gcs-bucket", "", "override default bucket")
+	var doProvisionalF = flag.Bool("provisional", false, "publish provisional binaries")
+	var doBlessF = flag.Bool("bless", false, "bless provisional binaries")
+
 	flag.Parse()
 	log.SetFlags(log.LstdFlags | log.Lshortfile)
 
+	var bucketName string
+	if len(*destBucket) > 0 {
+		bucketName = *destBucket
+	} else if *isReleaseF {
+		bucketName = "binaries.cockroachdb.com"
+	} else {
+		bucketName = "cockroach"
+	}
+	log.Printf("Using S3 bucket: %s", bucketName)
+
 	if _, ok := os.LookupEnv(awsAccessKeyIDKey); !ok {
 		log.Fatalf("AWS access key ID environment variable %s is not set", awsAccessKeyIDKey)
 	}
 	if _, ok := os.LookupEnv(awsSecretAccessKeyKey); !ok {
 		log.Fatalf("AWS secret access key environment variable %s is not set", awsSecretAccessKeyKey)
 	}
-	s3, err := makeS3()
+	var providers []release.ObjectPutGetter
+	s3, err := release.NewS3("us-east-1", bucketName)
 	if err != nil {
 		log.Fatalf("Creating AWS S3 session: %s", err)
 	}
+	providers = append(providers, s3)
+	if *gcsBucket != "" {
+		if _, ok := os.LookupEnv("GOOGLE_APPLICATION_CREDENTIALS"); !ok {
```

Comment:
Nit: this logic is duplicated. You can refactor this out and just put the validation in `release.NewGCS()`.

Reply:
Good point. Let me think if I want the function return an error or has the same signature as the S3 counterpart.

Reply:
I left it as is to keep the "panic" logic out of the "library", similarly to the S3 approach.